### DISCUSSION TITLE
How to create a patch from scratch

### DISCUSSION THREAD

**Discussion by bakkeby** [bakkeby/dwm-flexipatch#160] [How To]
Here is a step-by-step guide for creating patches from scratch.

**1)** clone the repository and cd into it

```
$ git clone git://git.suckless.org/dwm
Cloning into 'dwm'...
remote: Enumerating objects: 6404, done.
remote: Counting objects: 100% (6404/6404), done.
remote: Compressing objects: 100% (3116/3116), done.
remote: Total 6404 (delta 3668), reused 5933 (delta 3287), pack-reused 0
Receiving objects: 100% (6404/6404), 6.14 MiB | 1.57 MiB/s, done.
Resolving deltas: 100% (3668/3668), done.
```

```
$ cd dwm
```

**2)** check out v6.2 of dwm

This is if the patch is going to be for 6.2. I usually do this as that is a more consistent reference.

If your patch is to be for the latest master then skip this step.

```
$ git checkout 6.2
Note: switching to '6.2'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at cb3f58a Prepare 6.2 release.
```

**3)** create a branch for your patch

```
$ git branch segfault
$ git checkout segfault
Switched to branch 'segfault'
```
(branching and checking out can be done in one command, separated here for readability)

**4)** make your change

```
$ vi dwm.c
```

**5)** check that there are no unwanted changes before you commit

```diff
$ git diff

diff --git a/dwm.c b/dwm.c
index 4465af1..7fc236a 100644
--- a/dwm.c
+++ b/dwm.c
@@ -708,7 +708,8 @@ drawbar(Monitor *m)
                drw_text(drw, m->ww - sw, 0, sw, bh, 0, stext, 0);
        }

-       for (c = m->clients; c; c = c->next) {
+       /* This innovative change will make dwm segfault */
+       for (c = m->clients; c->next != c; c = c->next) {
                occ |= c->tags;
                if (c->isurgent)
                        urg |= c->tags;
```

**6)** commit your change

```
$ git commit -m "Awesome patch that makes dwm segfault" --all
[segfault b8781a1] Awesome patch that makes dwm segfault
 1 file changed, 2 insertions(+), 1 deletion(-)
```

**7)** create your patch

Here creating the patch against 6.2.

```
$ git format-patch --stdout 6.2 > dwm-segfault-6.2.diff
```

If you skipped step 2 and are creating the patch against latest master it would look like this:

```
$ git format-patch --stdout HEAD^ > dwm-segfault-20210621-67d76bd.diff
```
`20210621` here referring to the date the patch was last modified and `67d76bd` being the seven character git commit short hash of the branch point (i.e. the last commit on the master branch).

The short hash can be retrieved via this command:

```
$ git rev-parse --short master
67d76bd
```

Alternatively if you didn't create a branch etc. then you can create a patch for the very last commit, e.g.

```
$ git format-patch --stdout -1
```

The final file will look something like this:
```diff
$ cat dwm-segfault-6.2.diff
From b8781a182d64773354cd9af3e9a97cdd56a09417 Mon Sep 17 00:00:00 2001
From: bakkeby <bakkeby@gmail.com>
Date: Mon, 21 Jun 2021 09:54:50 +0200
Subject: [PATCH] Awesome patch that makes dwm segfault

---
 dwm.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dwm.c b/dwm.c
index 4465af1..7fc236a 100644
--- a/dwm.c
+++ b/dwm.c
@@ -708,7 +708,8 @@ drawbar(Monitor *m)
        drw_text(drw, m->ww - sw, 0, sw, bh, 0, stext, 0);
    }

-   for (c = m->clients; c; c = c->next) {
+   /* This innovative change will make dwm segfault */
+   for (c = m->clients; c->next != c; c = c->next) {
        occ |= c->tags;
        if (c->isurgent)
            urg |= c->tags;
--
2.31.1

```

You may also want to refer to https://suckless.org/hacking/.

**Comment by fitrh** [comment]
Sometimes i get a diff with mess indentation like this
```diff
 drawbar(Monitor *m)
 {
        int x, w, wdelta, tagscheme;
-        int y = barborder;
-        int h = bh - barborder * 2;
+        int y = m->bargap ? barborder : 0;
+        int h = bh - y * 2;
        int boxs = drw->fonts->h / 9;
        int boxw = drw->fonts->h / 6 + 2;
         unsigned int titlepad = lrpad / 2;

```
how to avoid this ? especially if using `expandtab` in vim a.k.a always using space

**Reply by bakkeby** [reply]
As a rule of thumb you should always use the same indentation that the surrounding code uses.

Given that the code here uses tabs for indentation you should probably not use expandtab (you could set noexpandtab instead). Alternatively you could add something in your ~/.vimrc that determines whether to use tabs or spaces depending on what is used in the file already \[[ref](https://unix.stackexchange.com/questions/63196/in-vim-how-can-i-automatically-determine-whether-to-use-spaces-or-tabs-for-inde)\].

While using spaces or tabs often comes down to personal preference I am only aware of the python programming officially favouring the use of spaces over tabs \[[ref](https://www.python.org/dev/peps/pep-0008/)\].

**Reply by fitrh** [reply]
I see, i just think that there are a way to avoid this without changing the whole indentation to space or tabs

**Comment by m-althaf** [comment]
Great I wanted to ask this
